tests/check_coding_style.sh

```diff
@@ -0,0 +1,169 @@
+# used to deal with such false positives and can also be used for allowed deviations.
+
+
+# Determine path of the current script
+HERE=$(cd `dirname $0` && pwd)
+SOURCES="${HERE}/../source"
+
+# Desired header, adapted for automatic parsing
+DESIRED_HEADER="/* FILENAME
+Copyright (c) 2xxx by NAME
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/"
+
+function die()
+{
+	echo "ERROR: $1"
+	exit 1
+}
+
+
+
+# This is a white-list of allowed exceptions. For now just included in the script;
+# might want to make this a separate text-file later.
+function filter_exceptions()
+{
+	grep -v \
+		-e "Files.cpp:		struct stat buf;" \
+		-e "Files.cpp:	struct _stat buf;" \
+		-e "Files.cpp:	struct stat buf;" \
+		-e "ImageBuffer.cpp:		struct jpeg_error_mgr jerr;" \
+		-e "Point.h:		struct {" \
+		-e "gl_header.h: #include <OpenGL/GL3.h>" \
+		-e "gl_header.h: #include <GL/glew.h>" \
+		-e "main.cpp: mismatch in header"
+}
+
+
+
+function report_issue()
+{
+	ERROR_MESSAGE="$1"
+	# Just forward pipe-in towards the filter
+	filter_exceptions | \
+	while read ISSUE; do
+		# Print the error-message once
+		if [ "${ERROR_MESSAGE}" != "" ]; then
+			echo ""
+			echo "******************************************************"
+			echo "${ERROR_MESSAGE}"
+			echo "******************************************************"
+			ERROR_MESSAGE=""
+		fi
+		echo "${ISSUE}"
+	done
+}
+
+
+
+function filter_non_matched_class()
+{
+	while read MATCH_LINE; do
+		FILE_NAME=$(echo "${MATCH_LINE}" | cut -d\: -f1 | sed "s/\.h$//")
+		CLASS_NAME=$(echo "${MATCH_LINE}" | cut -d\: -f2 | sed "s/^class //")
+		if [ "${FILE_NAME}" != "${CLASS_NAME}" ]; then
+			echo "${MATCH_LINE}"
+		fi
+	done
+}
+
+
+
+# Go to the sources directory
+cd "${SOURCES}" || die "Sources directory not found"
+
+
+
+# Files section
+grep -P -n "[^\x00-\x7F]" * |\
+	report_issue "Files: All files should be plain ASCII."
+
+find -type f | grep -v -e ".h$" -e ".cpp$" -e "WinApp.rc" | sed "s,^./,," |\
+	report_issue "Files: All C++ files should use \".h\" and \".cpp\" for their extensions."
+
+grep "^class" *.h | tr -d '\r' | grep -v -e ";$" | cut -d\: -f1,2 | sed "s,[{].*$,," | sed "s, *$,," | filter_non_matched_class |\
+	report_issue "Files: Each class \"MyClass\" should have its own .h and .cpp files, and they should be named \"MyClass.h\" and \"MyClass.cpp\""
+
+for FILE in *.{h,cpp}; do
+	HEADER=$(head -n 11 ${FILE})
+	HEADER_CHECK=$(echo "${HEADER}" | sed "s,^/\* ${FILE}$,/* FILENAME," |\
+		sed "s,^Copyright (c) 2[0-9]* by [A-Za-z0-9 ]*,Copyright (c) 2xxx by NAME,")
```

Comment:
This regex seems overly restrictive, e.g. it will flag any names with non-space punctuation.

I would prefer matching all "copyright by ..." lines and stripping them out entirely.

Reply:
Fixed by allowing any valid character except newlines in copyright holder names.